File: skincore_swift/SkinCore/Services/APIClient.swift
import Foundation

class APIClient {
    static let shared = APIClient()
    
    // TODO: Production'da gerçek URL'ye değiştir
    #if targetEnvironment(simulator)
    private let baseURL = "http://localhost:5192/api"
    #else
    private let baseURL = "http://192.168.0.15:5192/api"
    #endif
    
    private init() {}
    
    // MARK: - Generic Request
    
    func request<T: Codable>(
        endpoint: String,
        method: String = "GET",
        body: (any Encodable)? = nil,
        authenticated: Bool = false
    ) async throws -> T {
        guard let url = URL(string: "\(baseURL)\(endpoint)") else {
            throw APIClientError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        // Add JWT token if authenticated
        if authenticated, let token = KeychainService.shared.getAccessToken() {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        
        // Encode body
        if let body = body {
            request.httpBody = try JSONEncoder().encode(body)
        }
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIClientError.invalidResponse
        }
        
        guard (200...299).contains(httpResponse.statusCode) else {
            // Try to decode error message
            if let errorResponse = try? JSONDecoder().decode(MessageResponse.self, from: data) {
                throw APIClientError.serverError(errorResponse.message)
            }
            throw APIClientError.httpError(httpResponse.statusCode)
        }
        
        return try JSONDecoder().decode(T.self, from: data)
    }
    
    // MARK: - Auth Endpoints
    
    func register(email: String, password: String, fullName: String?) async throws -> MessageResponse {
        let body = RegisterRequest(email: email, password: password, fullName: fullName)
        return try await request(endpoint: "/auth/register", method: "POST", body: body)
    }
    
    func verifyEmail(email: String, code: String) async throws -> AuthResponse {
        let body = VerifyEmailRequest(email: email, code: code)
        return try await request(endpoint: "/auth/verify-email", method: "POST", body: body)
    }
    
    func login(email: String, password: String) async throws -> AuthResponse {
        let body = LoginRequest(email: email, password: password)
        return try await request(endpoint: "/auth/login", method: "POST", body: body)
    }
    
    func appleSignIn(identityToken: String, fullName: String?, email: String?) async throws -> AuthResponse {
        let body = AppleSignInRequestBody(identityToken: identityToken, fullName: fullName, email: email)
        return try await request(endpoint: "/auth/apple", method: "POST", body: body)
    }
    
    func resendCode(email: String) async throws -> MessageResponse {
        let body = ResendCodeRequest(email: email)
        return try await request(endpoint: "/auth/resend-code", method: "POST", body: body)
    }
    
    func refreshToken() async throws -> AuthResponse {
        guard let token = KeychainService.shared.getRefreshToken() else {
            throw APIClientError.noRefreshToken
        }
        let body = RefreshTokenRequest(refreshToken: token)
        return try await request(endpoint: "/auth/refresh", method: "POST", body: body)
    }
    
    func getMe() async throws -> UserResponse {
        return try await request(endpoint: "/auth/me", authenticated: true)
    }

    // MARK: - Product Endpoints

    func searchProductsByName(query: String, maxResults: Int = 5) async throws -> [Product] {
        let encodedQuery = query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? query
        return try await request(
            endpoint: "/products/search/name?query=\(encodedQuery)&maxResults=\(maxResults)"
        )
    }

    func getProductDetails(id: String) async throws -> ProductWithEnrichedIngredients {
        return try await request(endpoint: "/products/\(id)")
    }

    func searchProductsByImage(imageData: Data, ocrText: String?, maxResults: Int = 5) async throws -> [Product] {
        guard let url = URL(string: "\(baseURL)/products/search/image") else {
            throw APIClientError.invalidURL
        }

        let boundary = UUID().uuidString
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()

        // image field
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"image\"; filename=\"scan.jpg\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: image/jpeg\r\n\r\n".data(using: .utf8)!)
        body.append(imageData)
        body.append("\r\n".data(using: .utf8)!)

        // ocrText field
        if let ocrText = ocrText, !ocrText.isEmpty {
            body.append("--\(boundary)\r\n".data(using: .utf8)!)
            body.append("Content-Disposition: form-data; name=\"ocrText\"\r\n\r\n".data(using: .utf8)!)
            body.append(ocrText.data(using: .utf8)!)
            body.append("\r\n".data(using: .utf8)!)
        }

        // maxResults field
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"maxResults\"\r\n\r\n".data(using: .utf8)!)
        body.append("\(maxResults)".data(using: .utf8)!)
        body.append("\r\n".data(using: .utf8)!)

        body.append("--\(boundary)--\r\n".data(using: .utf8)!)
        request.httpBody = body

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw APIClientError.httpError((response as? HTTPURLResponse)?.statusCode ?? 0)
        }

        return try JSONDecoder().decode([Product].self, from: data)
    }

    // MARK: - Popular Endpoints

    func getPopularProducts(limit: Int = 10) async throws -> [PopularProductResponse] {
        return try await request(endpoint: "/popular?limit=\(limit)")
    }

    // MARK: - Search History Endpoints

    func getSearchHistory(limit: Int = 20) async throws -> [SearchHistoryResponse] {
        return try await request(endpoint: "/userprofile/search-history?limit=\(limit)", authenticated: true)
    }

    func addSearchHistory(_ req: AddSearchHistoryRequest) async throws -> MessageResponse {
        return try await request(endpoint: "/userprofile/search-history", method: "POST", body: req, authenticated: true)
    }

    func deleteSearchHistoryItem(itemId: String) async throws -> MessageResponse {
        return try await request(endpoint: "/userprofile/search-history/\(itemId)", method: "DELETE", authenticated: true)
    }

    func clearSearchHistory() async throws -> MessageResponse {
        return try await request(endpoint: "/userprofile/search-history", method: "DELETE", authenticated: true)
    }

    // MARK: - Favorites Endpoints

    func getFavorites() async throws -> [FavoriteResponse] {
        return try await request(endpoint: "/userprofile/favorites", authenticated: true)
    }

    func addFavorite(_ req: AddFavoriteRequest) async throws -> MessageResponse {
        return try await request(endpoint: "/userprofile/favorites", method: "POST", body: req, authenticated: true)
    }

    func removeFavorite(productId: String) async throws -> MessageResponse {
        return try await request(endpoint: "/userprofile/favorites/\(productId)", method: "DELETE", authenticated: true)
    }

    func toggleFavorite(_ req: AddFavoriteRequest) async throws -> ToggleFavoriteResponse {
        return try await request(endpoint: "/userprofile/favorites/toggle", method: "POST", body: req, authenticated: true)
    }

    func checkFavorite(productId: String) async throws -> IsFavoriteResponse {
        return try await request(endpoint: "/userprofile/favorites/\(productId)/check", authenticated: true)
    }

    // MARK: - Routines (Social) Endpoints

    func getRoutineFeed(limit: Int = 20, search: String? = nil) async throws -> [RoutineFeedItem] {
        var endpoint = "/routines?limit=\(limit)"
        if let search = search, !search.isEmpty {
            let encoded = search.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? search
            endpoint += "&search=\(encoded)"
        }
        return try await request(endpoint: endpoint, authenticated: true)
    }

    func getMyRoutines() async throws -> [RoutineFeedItem] {
        return try await request(endpoint: "/routines/my", authenticated: true)
    }

    func getRoutineDetail(id: String) async throws -> RoutineDetail {
        return try await request(endpoint: "/routines/\(id)", authenticated: true)
    }

    func createRoutine(_ req: CreateRoutineRequest) async throws -> RoutineFeedItem {
        return try await request(endpoint: "/routines", method: "POST", body: req, authenticated: true)
    }

    func addRoutineComment(routineId: String, text: String) async throws -> RoutineCommentResponse {
        let body = AddRoutineCommentRequest(text: text)
        return try await request(endpoint: "/routines/\(routineId)/comments", method: "POST", body: body, authenticated: true)
    }

    func toggleRoutineLike(routineId: String) async throws -> ToggleLikeResponse {
        return try await request(endpoint: "/routines/\(routineId)/likes/toggle", method: "POST", authenticated: true)
    }

    func uploadImage(data: Data) async throws -> String {
        guard let url = URL(string: "\(baseURL)/routines/upload-image") else {
            throw APIClientError.invalidURL
        }

        let boundary = UUID().uuidString
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        if let token = KeychainService.shared.getAccessToken() {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        var body = Data()
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"file\"; filename=\"photo.jpg\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: image/jpeg\r\n\r\n".data(using: .utf8)!)
        body.append(data)
        body.append("\r\n--\(boundary)--\r\n".data(using: .utf8)!)
        request.httpBody = body

        let (responseData, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw APIClientError.httpError((response as? HTTPURLResponse)?.statusCode ?? 0)
        }

        struct UploadResponse: Codable { let imageUrl: String }
        let result = try JSONDecoder().decode(UploadResponse.self, from: responseData)
        return result.imageUrl
    }

    // MARK: - Ingredients Endpoints

    func getIngredients(search: String? = nil, page: Int = 1, pageSize: Int = 50, minSafety: Int? = nil, maxSafety: Int? = nil, comedogenic: Bool? = nil) async throws -> [MatchedIngredient] {
        var endpoint = "/ingredients?page=\(page)&pageSize=\(pageSize)"
        if let search = search, !search.isEmpty {
            let encoded = search.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? search
            endpoint += "&search=\(encoded)"
        }
        if let min = minSafety {
            endpoint += "&minSafety=\(min)"
        }
        if let max = maxSafety {
            endpoint += "&maxSafety=\(max)"
        }
        if let com = comedogenic, com {
            endpoint += "&comedogenic=true"
        }
        return try await request(endpoint: endpoint)
    }
}

// MARK: - Errors

enum APIClientError: LocalizedError {
    case invalidURL
    case invalidResponse
    case httpError(Int)
    case serverError(String)
    case noRefreshToken
    
    var errorDescription: String? {
        switch self {
        case .invalidURL: return "Geçersiz URL"
        case .invalidResponse: return "Geçersiz yanıt"
        case .httpError(let code): return "HTTP hatası: \(code)"
        case .serverError(let message): return message
        case .noRefreshToken: return "Oturumunuz sona ermiş"
        }
    }
}

File: skincore_native/src/api/apiClient.ts
import axios, { AxiosInstance, InternalAxiosRequestConfig } from 'axios';
import * as SecureStore from 'expo-secure-store';
import type {
  AuthResponse,
  MessageResponse,
  UserResponse,
} from '../types/auth';
import type {
  Product,
  ProductWithEnrichedIngredients,
  PopularProductResponse,
  SearchHistoryResponse,
  AddSearchHistoryRequest,
  FavoriteResponse,
  AddFavoriteRequest,
  ToggleFavoriteResponse,
  IsFavoriteResponse,
  MatchedIngredient,
} from '../types/product';
import type {
  RoutineFeedItem,
  RoutineDetail,
  ToggleLikeResponse,
  RoutineCommentResponse,
  CreateRoutineRequest,
  UserProfileResponse,
  PublicUserProfileResponse,
  UpdateProfileRequest,
  UpdateBioRequest,
} from '../types/social';

// MARK: - Sabitler

// Localhost üzerinden .NET backend'e bağlanıyoruz
export const API_BASE_URL = 'http://localhost:5192/api';
export const MEDIA_BASE_URL = 'http://localhost:5192';

const ACCESS_TOKEN_KEY = 'com.skincore.accessToken';
const REFRESH_TOKEN_KEY = 'com.skincore.refreshToken';

// MARK: - Token Yönetimi (expo-secure-store → Swift KeychainService karşılığı)

export const TokenStorage = {
  async getAccessToken(): Promise<string | null> {
    return SecureStore.getItemAsync(ACCESS_TOKEN_KEY);
  },
  async getRefreshToken(): Promise<string | null> {
    return SecureStore.getItemAsync(REFRESH_TOKEN_KEY);
  },
  async saveTokens(response: AuthResponse): Promise<void> {
    await SecureStore.setItemAsync(ACCESS_TOKEN_KEY, response.accessToken);
    await SecureStore.setItemAsync(REFRESH_TOKEN_KEY, response.refreshToken);
  },
  async clearAll(): Promise<void> {
    await SecureStore.deleteItemAsync(ACCESS_TOKEN_KEY);
    await SecureStore.deleteItemAsync(REFRESH_TOKEN_KEY);
  },
};

// MARK: - Görsel URL Çözücü

export function resolveMediaUrl(path: string | undefined | null): string | undefined {
  if (!path || path.trim() === '') return undefined;
  if (path.startsWith('http')) return path;
  return `${MEDIA_BASE_URL}${path}`;
}

// MARK: - Axios Instance

const api: AxiosInstance = axios.create({
  baseURL: API_BASE_URL,
  timeout: 30000,
});

// Request interceptor: access token ekle
api.interceptors.request.use(async (config: InternalAxiosRequestConfig) => {
  const token = await TokenStorage.getAccessToken();
  if (token) {
    config.headers.Authorization = `Bearer ${token}`;
  }
  return config;
});

// Response interceptor: 401 → refresh token → retry
let isRefreshing = false;
let failedQueue: Array<{
  resolve: (token: string) => void;
  reject: (error: unknown) => void;
}> = [];

function processQueue(error: unknown, token: string | null) {
  failedQueue.forEach(({ resolve, reject }) => {
    if (error) reject(error);
    else if (token) resolve(token);
  });
  failedQueue = [];
}

api.interceptors.response.use(
  (response) => response,
  async (error) => {
    const originalRequest = error.config;

    if (error.response?.status === 401 && !originalRequest._retry) {
      if (isRefreshing) {
        return new Promise((resolve, reject) => {
          failedQueue.push({ resolve, reject });
        }).then((token) => {
          originalRequest.headers.Authorization = `Bearer ${token}`;
          return api(originalRequest);
        });
      }

      originalRequest._retry = true;
      isRefreshing = true;

      try {
        const refreshToken = await TokenStorage.getRefreshToken();
        if (!refreshToken) throw new Error('No refresh token');

        const response = await axios.post<AuthResponse>(
          `${API_BASE_URL}/auth/refresh`,
          { refreshToken }
        );
        await TokenStorage.saveTokens(response.data);
        const newToken = response.data.accessToken;
        processQueue(null, newToken);
        originalRequest.headers.Authorization = `Bearer ${newToken}`;
        return api(originalRequest);
      } catch (refreshError) {
        processQueue(refreshError, null);
        await TokenStorage.clearAll();
        // Auth store'u trigger etmek için event emit edilebilir, şimdilik throw
        return Promise.reject(refreshError);
      } finally {
        isRefreshing = false;
      }
    }

    return Promise.reject(error);
  }
);

// MARK: - Hata mesajı çıkarıcı

export function extractErrorMessage(error: unknown, fallback = 'Bir hata oluştu.'): string {
  console.error('[API Error]:', error);
  if (axios.isAxiosError(error)) {
    if (error.response?.data) {
      console.error('[API Error Data]:', error.response.data);
      if (error.response.data.message) {
        return error.response.data.message;
      }
      if (error.response.data.title) {
        return error.response.data.title;
      }
    }
    // Network Error veya timeout ise:
    return error.message ?? fallback;
  }
  if (error instanceof Error) return error.message;
  return fallback;
}

// MARK: - Auth Endpoints

export const authApi = {
  register: (body: { email: string; password: string; fullName?: string }) =>
    api.post<MessageResponse>('/auth/register', body).then((r) => r.data),

  verifyEmail: (body: { email: string; code: string }) =>
    api.post<AuthResponse>('/auth/verify-email', body).then((r) => r.data),

  login: (body: { email: string; password: string }) =>
    api.post<AuthResponse>('/auth/login', body).then((r) => r.data),

  appleSignIn: (body: { identityToken: string; fullName?: string; email?: string }) =>
    api.post<AuthResponse>('/auth/apple', body).then((r) => r.data),

  resendCode: (email: string) =>
    api.post<MessageResponse>('/auth/resend-code', { email }).then((r) => r.data),

  refreshToken: (refreshToken: string) =>
    api.post<AuthResponse>('/auth/refresh', { refreshToken }).then((r) => r.data),

  getMe: () => api.get<UserResponse>('/auth/me').then((r) => r.data),

  forgotPassword: (email: string) =>
    api.post<MessageResponse>('/auth/forgot-password', { email }).then((r) => r.data),

  resetPassword: (body: { email: string; code: string; newPassword: string }) =>
    api.post<MessageResponse>('/auth/reset-password', body).then((r) => r.data),

  changePassword: (body: { currentPassword: string; newPassword: string }) =>
    api.put<MessageResponse>('/auth/change-password', body).then((r) => r.data),

  updateNotifications: (enabled: boolean) =>
    api.put<MessageResponse>('/auth/notifications', { enabled }).then((r) => r.data),

  deleteAccount: () =>
    api.delete<MessageResponse>('/auth/account').then((r) => r.data),
};

// MARK: - Product Endpoints

export const productsApi = {
  searchByName: (query: string, maxResults = 5) =>
    api
      .get<Product[]>('/products/search/name', { params: { query, maxResults } })
      .then((r) => r.data),

  getDetails: (id: string) =>
    api.get<ProductWithEnrichedIngredients>(`/products/${id}`).then((r) => r.data),

  searchByImage: async (
    imageData: FormData
  ): Promise<Product[]> => {
    const response = await api.post<Product[]>('/products/search/image', imageData);
    return response.data;
  },

  submitRequest: async (formData: FormData): Promise<MessageResponse> => {
    const response = await api.post<MessageResponse>('/product-requests', formData);
    return response.data;
  },
};

// MARK: - Moderation Endpoints

export const moderationApi = {
  submitRequest: async (productId: string, formData: FormData): Promise<MessageResponse> => {
    const response = await api.post<MessageResponse>(
      `/moderation/products/${productId}`,
      formData
    );
    return response.data;
  },
};

// MARK: - Popular Endpoints

export const popularApi = {
  get: (limit = 10) =>
    api.get<PopularProductResponse[]>('/popular', { params: { limit } }).then((r) => r.data),
};

// MARK: - Search History Endpoints

export const searchHistoryApi = {
  get: (limit = 20) =>
    api
      .get<SearchHistoryResponse[]>('/userprofile/search-history', { params: { limit } })
      .then((r) => r.data),

  add: (body: AddSearchHistoryRequest) =>
    api
      .post<MessageResponse>('/userprofile/search-history', body)
      .then((r) => r.data),

  deleteItem: (itemId: string) =>
    api
      .delete<MessageResponse>(`/userprofile/search-history/${itemId}`)
      .then((r) => r.data),

  clearAll: () =>
    api.delete<MessageResponse>('/userprofile/search-history').then((r) => r.data),
};

// MARK: - Favorites Endpoints

export const favoritesApi = {
  get: () =>
    api.get<FavoriteResponse[]>('/userprofile/favorites').then((r) => r.data),

  add: (body: AddFavoriteRequest) =>
    api.post<MessageResponse>('/userprofile/favorites', body).then((r) => r.data),

  remove: (productId: string) =>
    api.delete<MessageResponse>(`/userprofile/favorites/${productId}`).then((r) => r.data),

  toggle: (body: AddFavoriteRequest) =>
    api
      .post<ToggleFavoriteResponse>('/userprofile/favorites/toggle', body)
      .then((r) => r.data),

  check: (productId: string) =>
    api
      .get<IsFavoriteResponse>(`/userprofile/favorites/${productId}/check`)
      .then((r) => r.data),
};

// MARK: - Routines Endpoints

export const routinesApi = {
  getFeed: (limit = 20, search?: string) =>
    api
      .get<RoutineFeedItem[]>('/routines', { params: { limit, ...(search ? { search } : {}) } })
      .then((r) => r.data),

  getMy: () => api.get<RoutineFeedItem[]>('/routines/my').then((r) => r.data),

  getFavorites: (limit = 20) =>
    api
      .get<RoutineFeedItem[]>('/routines/favorites', { params: { limit } })
      .then((r) => r.data),

  getByUserId: (userId: string) =>
    api.get<RoutineFeedItem[]>(`/routines/user/${userId}`).then((r) => r.data),

  getLikedByUserId: (userId: string) =>
    api.get<RoutineFeedItem[]>(`/routines/user/${userId}/favorites`).then((r) => r.data),

  getDetail: (id: string) =>
    api.get<RoutineDetail>(`/routines/${id}`).then((r) => r.data),

  create: (body: CreateRoutineRequest) =>
    api.post<RoutineFeedItem>('/routines', body).then((r) => r.data),

  delete: (routineId: string) =>
    api.delete<MessageResponse>(`/routines/${routineId}`).then((r) => r.data),

  addComment: (routineId: string, text: string) =>
    api
      .post<RoutineCommentResponse>(`/routines/${routineId}/comments`, { text })
      .then((r) => r.data),

  toggleLike: (routineId: string) =>
    api
      .post<ToggleLikeResponse>(`/routines/${routineId}/likes/toggle`)
      .then((r) => r.data),

  uploadImage: async (formData: FormData): Promise<string> => {
    const response = await api.post<{ imageUrl: string }>(
      '/routines/upload-image',
      formData
    );
    return response.data.imageUrl;
  },
};

// MARK: - Profile Endpoints

export const profileApi = {
  uploadImage: async (formData: FormData): Promise<string> => {
    const response = await api.post<{ imageUrl: string }>(
      '/userprofile/profile-image',
      formData
    );
    return response.data.imageUrl;
  },

  update: (body: UpdateProfileRequest) =>
    api.put<MessageResponse>('/userprofile', body).then((r) => r.data),

  checkUsername: async (username: string): Promise<boolean> => {
    const res = await api.post<{ isAvailable: boolean }>(
      '/userprofile/check-username',
      { username }
    );
    return res.data.isAvailable;
  },

  getMyProfile: () =>
    api.get<UserProfileResponse>('/userprofile').then((r) => r.data),

  updateBio: (body: UpdateBioRequest) =>
    api.put<MessageResponse>('/userprofile/bio', body).then((r) => r.data),

  follow: (userId: string) =>
    api.post<MessageResponse>(`/userprofile/follow/${userId}`).then((r) => r.data),

  unfollow: (userId: string) =>
    api.delete<MessageResponse>(`/userprofile/follow/${userId}`).then((r) => r.data),

  getFollowers: () =>
    api.get<PublicUserProfileResponse[]>('/userprofile/followers').then((r) => r.data),

  getFollowing: () =>
    api.get<PublicUserProfileResponse[]>('/userprofile/following').then((r) => r.data),

  searchUsers: (query: string) =>
    api
      .get<PublicUserProfileResponse[]>('/userprofile/search', {
        params: { query, limit: 15 },
      })
      .then((r) => r.data),

  getPublicProfile: (username: string) =>
    api
      .get<PublicUserProfileResponse>(`/userprofile/public/${encodeURIComponent(username)}`)
      .then((r) => r.data),

  getPublicFollowers: (username: string) =>
    api
      .get<PublicUserProfileResponse[]>(
        `/userprofile/public/${encodeURIComponent(username)}/followers`
      )
      .then((r) => r.data),

  getPublicFollowing: (username: string) =>
    api
      .get<PublicUserProfileResponse[]>(
        `/userprofile/public/${encodeURIComponent(username)}/following`
      )
      .then((r) => r.data),

  getPublicFavorites: (username: string) =>
    api
      .get<FavoriteResponse[]>(
        `/userprofile/public/${encodeURIComponent(username)}/favorites`
      )
      .then((r) => r.data),
};

// MARK: - Ingredients Endpoints

export interface IngredientsFilters {
  search?: string;
  page?: number;
  pageSize?: number;
  minSafety?: number;
  maxSafety?: number;
  comedogenic?: boolean;
  safetyLabel?: string;
}

export const ingredientsApi = {
  get: (filters: IngredientsFilters = {}) => {
    const params: Record<string, string | number | boolean> = {
      page: filters.page ?? 1,
      pageSize: filters.pageSize ?? 50,
    };
    if (filters.search) params.search = filters.search;
    if (filters.minSafety !== undefined) params.minSafety = filters.minSafety;
    if (filters.maxSafety !== undefined) params.maxSafety = filters.maxSafety;
    if (filters.comedogenic) params.comedogenic = true;
    if (filters.safetyLabel) params.safetyLabel = filters.safetyLabel;

    return api.get<MatchedIngredient[]>('/ingredients', { params }).then((r) => r.data);
  },
};

export default api;
